engine/metrics.go

```diff
@@ -328,16 +329,22 @@ func (m *MetricsClient) Send() error {
 	if m.statsd == "" {
 		return nil
 	}
-	remote, err := statsdlib.New(m.statsd)
-	if err != nil {
-		log.Errorf("[statsd] Connect statsd failed: %v", err)
-		return err
+	//Lazy connecting
+	if m.statsdClient == nil {
+		// We needn't try to renew/reconnect because of only supporting UDP protocol now
+		// We should add an `errorCount` to reconnect when implementing TCP protocol
+		client, err := statsdlib.New(m.statsd, statsdlib.WithErrorHandler(func (err error) {
+			log.Errorf("[statsd] Sending statsd failed: %v", err)
+		}))
+		if err != nil {
+			log.Errorf("[statsd] Connect statsd failed: %v", err)
+			return err
+		}
+		m.statsdClient = client
```

Comment:
```
	if m.statsdClient != nil {
		return nil
	}
	var err error
	// We needn't try to renew/reconnect because of only supporting UDP protocol now
	// We should add an `errorCount` to reconnect when implementing TCP protocol
	if m.statsdClient, err = statsdlib.New(m.StatsdAddr, statsdlib.WithErrorHandler(func(err error) {
		log.Errorf("[statsd] Sending statsd failed: %v", err)
	})); err != nil {
		log.Errorf("[statsd] Connect statsd failed: %v", err)
		return err
	}
```

Reply:
```go
if m.statsdClient != nil {
		return nil
}
```
在``` == nil``` 块内又判断一次是为什么？这里貌似没并发，有的话应该加锁实现

用c的时候一般大家会约定尽量别在if里写太多，在我看来go还是更像c一点，那我交换下写法
另外这里本来新建一个局部变量是减小临界区且避免失败调用时影响实例变量，如果考虑上无并发的场景是可以直接写成员变量

Reply:
我看上下文逻辑是对 m.statsdClient 做是否为空判断吧，如果为空就重连来着……所以用了个前置 if 来提早分支结束流程，然后在后面直接写成员变量。看这里的用况应该是安全的。